src/code/padmgr.c

```diff
@@ -23,235 +84,260 @@ OSMesgQueue* PadMgr_LockSerialMesgQueue(PadMgr* padMgr) {
+                // "Stop rumble pak"
+                PADMGR_LOG(i, "振動パック 停止");
+            }
+            osMotorStop(&padMgr->rumblePfs[i]);
         }
     }
 
-    PadMgr_UnlockSerialMesgQueue(padMgr, serialEventQueue);
+    PadMgr_ReleaseSerialEventQueue(padMgr, serialEventQueue);
 }
 
+/**
+ * Prevents rumble for 3 VI, ~0.05 seconds at 60 VI/sec
+ */
 void PadMgr_RumbleReset(PadMgr* padMgr) {
-    padMgr->rumbleOffFrames = 3;
+    padMgr->rumbleOffTimer = 3;
 }
 
-void PadMgr_RumbleSetSingle(PadMgr* padMgr, u32 ctrlr, u32 rumble) {
-    padMgr->rumbleEnable[ctrlr] = rumble;
-    padMgr->rumbleOnFrames = 240;
+/**
+ * Enables or disables rumble on controller port `port` for 240 VI,
+ * ~4 seconds at 60 VI/sec and ~4.8 seconds at 50 VI/sec
+ */
+void PadMgr_RumbleSetSingle(PadMgr* padMgr, u32 port, u32 rumble) {
+    padMgr->rumbleEnable[port] = rumble;
+    padMgr->rumbleOnTimer = 240;
 }
 
-void PadMgr_RumbleSet(PadMgr* padMgr, u8* ctrlrRumbles) {
+/**
+ * Enables or disables rumble on all controller ports for 240 VI,
+ * ~4 seconds at 60 VI/sec and ~4.8 seconds at 50 VI/sec
+ *
+ * @param enable    Array of 4 u8 containing either true or false to enable or disable rumble for that controller
+ */
+void PadMgr_RumbleSet(PadMgr* padMgr, u8 enable[MAXCONTROLLERS]) {
     s32 i;
 
-    for (i = 0; i < 4; i++) {
-        padMgr->rumbleEnable[i] = ctrlrRumbles[i];
+    for (i = 0; i < MAXCONTROLLERS; i++) {
+        padMgr->rumbleEnable[i] = enable[i];
     }
 
-    padMgr->rumbleOnFrames = 240;
+    padMgr->rumbleOnTimer = 240;
 }
 
-void PadMgr_ProcessInputs(PadMgr* padMgr) {
+/**
+ * Updates `padMgr->inputs` based on the error response of each controller
+ */
+void PadMgr_UpdateInputs(PadMgr* padMgr) {
     s32 i;
     Input* input;
-    OSContPad* padnow1; // original name
+    OSContPad* pad; // original name: "padnow1"
     s32 buttonDiff;
 
     PadMgr_LockPadData(padMgr);
 
-    input = &padMgr->inputs[0];
-    padnow1 = &padMgr->pads[0];
-
-    for (i = 0; i < padMgr->nControllers; i++, input++, padnow1++) {
+    for (input = &padMgr->inputs[0], pad = &padMgr->pads[0], i = 0; i < padMgr->nControllers; i++, input++, pad++) {
         input->prev = input->cur;
 
-        if (1) {} // Necessary to match
-
-        switch (padnow1->errno) {
+        switch (pad->errno) {
             case 0:
-                input->cur = *padnow1;
+                // No error, copy inputs
+                input->cur = *pad;
                 if (!padMgr->ctrlrIsConnected[i]) {
                     padMgr->ctrlrIsConnected[i] = true;
-                    osSyncPrintf(VT_FGCOL(YELLOW));
-                    osSyncPrintf("padmgr: %dコン: %s\n", i + 1, "認識しました"); // "Recognized"
-                    osSyncPrintf(VT_RST);
+                    // "Recognized"
+                    PADMGR_LOG(i, "認識しました");
                 }
                 break;
-            case 4:
+            case (CHNL_ERR_OVERRUN >> 4):
```

Comment:
where do these defines come from?
os_cont.h http://n64devkit.square7.ch/header/os_cont.htm has
```c
/* controller errors */
#define CONT_NO_RESPONSE_ERROR          0x8
#define CONT_OVERRUN_ERROR              0x4
#ifdef _HW_VERSION_1
#define CONT_FRAME_ERROR                0x2
#define CONT_COLLISION_ERROR            0x1
#endif 
```
but if it wasn't for seeing this (`CHNL_ERR_OVERRUN`) I would have never tied those defines to `OSContPad.errno`

Reply:
It's one of the secrets of rcp.h

Reply:
I swear I grep'ed and didn't find it before. found it now thanks :sweat_smile: